# IMPORTACIÓN DE LIBRERÍAS

In [1]:
import pandas as pd

# IMPORTACIÓN DEL DATASET

In [2]:
!curl -L -o telefonos-2024.zip\
https://www.kaggle.com/api/v1/datasets/download/jakubkhalponiak/phones-2024

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1352k  100 1352k    0     0  1625k      0 --:--:-- --:--:-- --:--:-- 9593k


In [3]:
!unzip telefonos-2024.zip

Archive:  telefonos-2024.zip
  inflating: data.json               
  inflating: processed_data2.csv     


In [4]:
df = pd.read_csv('processed_data2.csv')
df_J = pd.read_json(('data.json'))

In [5]:
!head processed_data2.csv

phone_brand,phone_model,store,price_usd,storage,ram,launch_date,dimensions,weight,display_type,display_size,display_resolution,os,nfc,usb,battery,features_sensors,colors,video,chipset,cpu,gpu,year,foldable,ppi_density,quantile_10,quantile_50,quantile_90,price_range,os_type,os_version,battery_size,colors_available,chip_company,cpu_core,gpu_company,fingerprint,video_resolution
apple,Apple iPhone 16 Pro,Amazon DE,1357.55,256,8,2024-09-20,149.6 x 71.5 x 8.3 mm (5.89 x 2.81 x 0.33 in),199.0,"LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolby Vision, 1000 nits (typ), 2000 nits (HBM)",6.3,1206 x 2622,iOS 18,1,"USB Type-C 3.2 Gen 2, DisplayPort",3582,"Face ID, accelerometer, gyro, proximity, compass, barometer","Black Titanium, White Titanium, Natural Titanium, Desert Titanium","4K@24/25/30/60/100/120fps, 1080p@25/30/60/120/240fps, 10-bit HDR, Dolby Vision HDR (up to 60fps), ProRes, 3D (spatial) video/audio, stereo sound rec.",Apple A18 Pro (3 nm),Hexa-core (2x4.05 GHz + 4x2.42 GHz),Apple GPU (6-

In [6]:
!head data.json

[
{"phone_brand": "itel", "phone_model": "itel Smart Watch 1", "price": null, "specs": {"Network": {"2G bands": " N/A", "3G bands": " N/A", "4G bands": " N/A", "Speed": "No", "GPRS": "No", "EDGE": "No"}, "Launch": {"Announced": "2022", "Status": "Available. Released 2022"}, "Body": {"Dimensions": "-", "Weight": "-", "Build": "Plastic back, plastic frame", "SIM": "No"}, "Display": {"Type": "IPS LCD", "Size": "1.72 inches", "Resolution": "240 x 283 pixels (~216 ppi density)"}, "Platform": {"OS": "Proprietary OS"}, "Memory": {"Card slot": "No", "Internal": "Unspecified storage"}, "Sound": {"Loudspeaker": "Yes", "3.5mm jack": "No"}, "Comms": {"WLAN": "No", "Bluetooth": "5.0, A2DP, LE", "Positioning": "No", "NFC": "No", "Radio": "No", "USB": "No"}, "Features": {"Sensors": "Accelerometer, heart rate, SpO2"}, "Battery": {"Type": "200 mAh, non-removable"}, "Misc": {"Colors": "Black"}}},
{"phone_brand": "oukitel", "phone_model": "Oukitel WP19", "price": "About 380 EUR", "specs": {"Network": {"2

# DESCRIPCIÓN DEL DATASET

El dataset cuenta con 2 conjuntos de datos separados en 2 archivos data.json y processed_data2.csv relacionados por la marca y modelo del teléfono.

Archivo data.json:

Este archivo contiene los datos sin procesar extraídos de GSMArena.com, incluida información sobre varios teléfonos actualmente disponibles en el mercado.


- phone_brand: El fabricante o la marca del teléfono (por ejemplo, Apple, Samsung, Xiaomi).
- phone_model: El nombre o número de modelo específico del teléfono.
- Precio: El precio del teléfono, que puede ser una cifra exacta o una estimación aproximada. Nota: Los datos de precios pueden requerir una revisión.
- especificaciones: Un diccionario de diccionarios que detalla las especificaciones técnicas del teléfono, incluidas características como el tamaño de la pantalla, la resolución de la cámara, el procesador, la duración de la batería y más.
- precios: Un diccionario de diccionarios que contiene listas de precios del teléfono en varios sitios web de comercio electrónico.

In [7]:
df_J.head()

phone_brand         phone_model          price  \
0        itel  itel Smart Watch 1           None   
1     oukitel        Oukitel WP19  About 380 EUR   
2       cubot   Cubot Smart Watch           None   
3       cubot         Cubot ID206           None   
4         tcl            TCL Plex  About 330 EUR   

                                               specs pricing  
0  {'Network': {'2G bands': ' N/A', '3G bands': '...     NaN  
1  {'Network': {'2G bands': 'GSM 850 / 900 / 1800...     NaN  
2  {'Network': {'2G bands': ' N/A', '3G bands': '...     NaN  
3  {'Network': {'2G bands': ' N/A', '3G bands': '...     NaN  
4  {'Network': {'2G bands': 'GSM 850 / 900 / 1800...     NaN

Archivo processed_data2.csv:

Este archivo contiene datos limpios y procesado sobre teléfonos, con información de varias plataformas de comercio electrónico.

- phone_brand: El fabricante o la marca del teléfono (por ejemplo, Apple, Samsung, Xiaomi).
- phone_model: El nombre o número de modelo específico del teléfono.
- tienda: La tienda específica donde se encuentra listado el teléfono.
- precio: el precio del teléfono en forma de número de punto flotante en la moneda local. Esta columna siempre debe analizarse junto con la currencycolumna
- Moneda: La moneda en la que se establece el precio (por ejemplo, USD, EUR).
- price_USD: El precio del teléfono convertido a USD.
- almacenamiento: La cantidad de almacenamiento en GB (entero).
- ram: La cantidad de RAM en GB (entero).
- Lanzamiento: La fecha de lanzamiento del teléfono (en formato de fecha y hora).
- Dimensiones: Las dimensiones exactas del teléfono, por ejemplo: 163.8 x 76.8 x 8.9 mm (6.45 x 3.02 x 0.35 in).
- Peso: El peso del teléfono en gramos (flotante).
- Display_Type: El tipo de visualización, por ejemplo: LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolby Vision, 1000 nits (typ), 2000 nits (HBM).
- Display_Size: El tamaño de la pantalla en pulgadas.
- Display_Resolution: La resolución de la pantalla, por ejemplo: 1280 x 2856.
- SO: El sistema operativo del teléfono, por ejemplo: iOS 18, Android 14.
- NFC: Una bandera que indica si el teléfono tiene NFC: 1- Tiene NFC, 0- No tiene NFC.
- USB: El tipo de USB, por ejemplo: USB Type-C 3.2 Gen 2, DisplayPort.
- BATERÍA: La capacidad de la batería en mAh.
- Características_Sensores: Características y sensores incluidos en el teléfono.
- Colores: Las opciones de color para el modelo de teléfono, por ejemplo: Black Titanium, White Titanium, Natural Titanium, Desert Titanium.
- Vídeo: Especificaciones de la cámara para grabación de vídeo, por ejemplo:4K@24/25/30/60/100/120fps, 1080p@25/30/60/120/240fps, 10-bit HDR, Dolby Vision HDR (up to 60fps), ProRes, 3D (spatial) video/audio, stereo sound rec.
- Chipset: El chipset utilizado en el teléfono, por ejemplo: Apple A18 Pro (3 nm).
- CPU: Las especificaciones de la CPU, por ejemplo: Hexa-core (2x4.05 GHz + 4x2.42 GHz).
- GPU: Las especificaciones de la GPU, por ejemplo: Apple GPU (6-core graphics).
- Año: El año en el que se lanzó el modelo de teléfono.
- Plegable: Una bandera que indica si el teléfono es plegable: 1- El teléfono es plegable, 0- El teléfono no es plegable.
- PPI_Density: La densidad de píxeles en píxeles por pulgada (ppi).
- quantile_10: El décimo cuartil del precio de los teléfonos en un año determinado.
- quantile_50: El 50.º cuartil (mediana) del precio de los teléfonos en un año determinado.
- quantile_90: El cuartil 90 del precio de los teléfonos en un año determinado.
- price_range: Clasificación del rango de precios ( low price, medium price, high price), basada en cuantiles en un año determinado.


In [8]:
df.head()

phone_brand              phone_model      store  price_usd  storage  ram  \
0       apple      Apple iPhone 16 Pro  Amazon DE    1357.55      256    8   
1       apple      Apple iPhone 16 Pro  Amazon DE    1492.55      512    8   
2       apple      Apple iPhone 16 Pro  Amazon DE    1705.32     1000    8   
3       apple  Apple iPhone 16 Pro Max  Amazon DE    1564.92      512    8   
4       apple     Apple iPhone 12 mini  Amazon DE     247.32      128    4   

  launch_date                                     dimensions  weight  \
0  2024-09-20  149.6 x 71.5 x 8.3 mm (5.89 x 2.81 x 0.33 in)   199.0   
1  2024-09-20  149.6 x 71.5 x 8.3 mm (5.89 x 2.81 x 0.33 in)   199.0   
2  2024-09-20  149.6 x 71.5 x 8.3 mm (5.89 x 2.81 x 0.33 in)   199.0   
3  2024-09-20    163 x 77.6 x 8.3 mm (6.42 x 3.06 x 0.33 in)   227.0   
4  2020-11-13  131.5 x 64.2 x 7.4 mm (5.18 x 2.53 x 0.29 in)   135.0   

                                        display_type  ...   price_range  \
0  LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolb...  ...  medium price   
1  LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolb...  ...    high price   
2  LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolb...  ...    high price   
3  LTPO Super Retina XDR OLED, 120Hz, HDR10, Dolb...  ...    high price   
4  Super Retina XDR OLED, HDR10, Dolby Vision, 62...  ...  medium price   

  os_type os_version  battery_size colors_available  chip_company   cpu_core  \
0     iOS         18        Medium                4         Apple  Hexa-core   
1     iOS         18        Medium                4         Apple  Hexa-core   
2     iOS         18        Medium                4         Apple  Hexa-core   
3     iOS         18         Large                4         Apple  Hexa-core   
4     iOS       14.1         Small                6         Apple  Hexa-core   

  gpu_company fingerprint video_resolution  
0       Apple        Face               4K  
1       Apple        Face               4K  
2       Apple        Face               4K  
3       Apple        Face               4K  
4       Apple        Face               4K  

[5 rows x 38 columns]

# Copiar archivos al sistema HDFS

## Instalación de Hadoop

El primer paso será instalar Hadoop, este framework nos montará un sistema de archivos HDFS

Nos descargamos el paquete instalador

In [9]:
!wget https://downloads.apache.org/hadoop/common/hadoop-3.3.6/hadoop-3.3.6.tar.gz

--2024-12-19 11:56:35--  https://downloads.apache.org/hadoop/common/hadoop-3.3.6/hadoop-3.3.6.tar.gz
Resolving downloads.apache.org (downloads.apache.org)... 88.99.208.237, 135.181.214.104, 2a01:4f8:10a:39da::2, ...
Connecting to downloads.apache.org (downloads.apache.org)|88.99.208.237|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 730107476 (696M) [application/x-gzip]
Saving to: ‘hadoop-3.3.6.tar.gz’

hadoop-3.3.6.tar.gz 100%[===================>] 696.28M  27.6MB/s    in 25s     

2024-12-19 11:57:01 (27.8 MB/s) - ‘hadoop-3.3.6.tar.gz’ saved [730107476/730107476]



Descomprimimos el contenido del archivo descargado

In [10]:
!tar -xzf hadoop-3.3.6.tar.gz

Una vez tenemos el contenido descomprimido, eliminamos el archivo descargado

In [11]:
!rm hadoop-3.3.6.tar.gz

Movemos la carpeta descomprimida a la ruta /usr/local

In [12]:
!mv  hadoop-3.3.6/ /usr/local/

Actuzlizamos al path del sistema para poder utilizar hadoop

In [13]:
import os
os.environ["JAVA_HOME"]="/usr/lib/jvm/java-11-openjdk-amd64"
os.environ["PATH"] = os.environ["PATH"] + ":" + "/usr/local/hadoop-3.3.6/bin"

Comprobamos que ya podemos ejecutar hadoop

In [14]:
!hadoop version

Hadoop 3.3.6
Source code repository https://github.com/apache/hadoop.git -r 1be78238728da9266a4f88195058f08fd012bf9c
Compiled by ubuntu on 2023-06-18T08:22Z
Compiled on platform linux-x86_64
Compiled with protoc 3.7.1
From source with checksum 5652179ad55f76cb287d9c633bb53bbd
This command was run using /usr/local/hadoop-3.3.6/share/hadoop/common/hadoop-common-3.3.6.jar


## Subir los archivos a hdfs

Creamos una carpeta en el entorno de hdfs

In [15]:
!hdfs dfs -mkdir /dataset

Subimos ambos archivos a la carpeta creada dentro de hdfs

In [16]:
!hdfs dfs -put processed_data2.csv /dataset

In [17]:
!hdfs dfs -put data.json /dataset

Comprobamos que los 2 archivos están subidos

In [18]:
!hdfs dfs -ls /dataset

Found 2 items
-rw-r--r--   1 root root   14322766 2024-12-19 11:57 /dataset/data.json
-rw-r--r--   1 root root    1082761 2024-12-19 11:57 /dataset/processed_data2.csv


# Conteo mediante grep

Analizaremos el arhivo processed_data2.csv y contaremos las veces que aparece la palabra Apple y la palabra Samsung y así podemos ver cuantos modelos de cada marca hay en el dataset

Guardamos el resultado en hdfs

In [19]:
%%bash
hadoop jar /usr/local/hadoop-3.3.6/share/hadoop/mapreduce/hadoop-mapreduce-examples-3.3.6.jar grep /dataset grep_apple 'apple'

2024-12-19 11:57:29,254 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2024-12-19 11:57:29,535 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2024-12-19 11:57:29,536 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2024-12-19 11:57:29,912 INFO input.FileInputFormat: Total input files to process : 2
2024-12-19 11:57:29,963 INFO mapreduce.JobSubmitter: number of splits:2
2024-12-19 11:57:30,573 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local691720336_0001
2024-12-19 11:57:30,573 INFO mapreduce.JobSubmitter: Executing with tokens: []
2024-12-19 11:57:31,063 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
2024-12-19 11:57:31,064 INFO mapreduce.Job: Running job: job_local691720336_0001
2024-12-19 11:57:31,115 INFO mapred.LocalJobRunner: OutputCommitter set in config null
2024-12-19 11:57:31,132 INFO output.PathOutputCommitterFactory: No output committer factory defined, defaultin

In [20]:
!hdfs dfs -ls grep_apple

Found 2 items
-rw-r--r--   1 root root          0 2024-12-19 11:57 grep_apple/_SUCCESS
-rw-r--r--   1 root root         10 2024-12-19 11:57 grep_apple/part-r-00000


In [21]:
!hdfs dfs -cat grep_apple/part-r-00000

322	apple


Guardamos el resultado en local

In [22]:
%%bash
hadoop jar /usr/local/hadoop-3.3.6/share/hadoop/mapreduce/hadoop-mapreduce-examples-3.3.6.jar grep /dataset /content/samsung 'samsung'

2024-12-19 11:57:40,721 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2024-12-19 11:57:40,913 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2024-12-19 11:57:40,913 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2024-12-19 11:57:41,248 INFO input.FileInputFormat: Total input files to process : 2
2024-12-19 11:57:41,295 INFO mapreduce.JobSubmitter: number of splits:2
2024-12-19 11:57:41,606 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local1255638818_0001
2024-12-19 11:57:41,606 INFO mapreduce.JobSubmitter: Executing with tokens: []
2024-12-19 11:57:41,884 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
2024-12-19 11:57:41,885 INFO mapreduce.Job: Running job: job_local1255638818_0001
2024-12-19 11:57:41,894 INFO mapred.LocalJobRunner: OutputCommitter set in config null
2024-12-19 11:57:41,903 INFO output.PathOutputCommitterFactory: No output committer factory defined, default

In [23]:
!ls -l /content/samsung

total 4
-rw-r--r-- 1 root root 13 Dec 19 11:57 part-r-00000
-rw-r--r-- 1 root root  0 Dec 19 11:57 _SUCCESS


In [24]:
!cat ./samsung/part-r-00000

1828	samsung


Vemos que la palabra apple aparece 322 veces y la palabra samsung aparece 1828 veces

### Conclusión

Podemos ver que aunque son las 2 marcas más conocidas por todos, samsung necesita diseñar y poner en el mercado 5 veces mas modelos distintos de móviles para poder competir con Apple.

# MarReduce en Java

Generamos el código .java

In [25]:
%%writefile -a wordcount_mapreduce.java

import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

public class wordcount_mapreduce {

    public static class WordCountMapper extends Mapper<Object, Text, Text, IntWritable> {

        private Text wordToCount = new Text();
        private final static IntWritable one = new IntWritable(1);

        // Expresión regular para buscar la palabra objetivo
        private Pattern targetPattern;

        @Override
        protected void setup(Context context) throws IOException, InterruptedException {
            // Obtenemos la palabra a buscar del contexto
            Configuration conf = context.getConfiguration();
            String targetWord = conf.get("targetWord").toLowerCase();

            // Creamos un patrón que busca la palabra completa (\b asegura límites de palabra)
            targetPattern = Pattern.compile("\\b" + Pattern.quote(targetWord) + "\\b", Pattern.CASE_INSENSITIVE);
        }

        @Override
        public void map(Object key, Text value, Context context) throws IOException, InterruptedException {
            // Obtener el nombre del archivo actual
            String filename = ((org.apache.hadoop.mapreduce.lib.input.FileSplit) context.getInputSplit()).getPath().getName();

            // Convertimos la línea a minúsculas
            String line = value.toString().toLowerCase();

            // Usamos el patrón para buscar coincidencias en la línea
            Matcher matcher = targetPattern.matcher(line);
            while (matcher.find()) {
                // Si encontramos la palabra, emitimos (archivo_nombre : palabra, 1)
                wordToCount.set(filename + " : " + matcher.group());
                context.write(wordToCount, one);
            }
        }
    }

    public static class WordCountReducer extends Reducer<Text, IntWritable, Text, IntWritable> {

        private IntWritable result = new IntWritable();

        @Override
        public void reduce(Text key, Iterable<IntWritable> values, Context context) throws IOException, InterruptedException {
            int sum = 0;

            // Sumamos todas las ocurrencias de la palabra
            for (IntWritable val : values) {
                sum += val.get();
            }

            result.set(sum);
            context.write(key, result);
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 3) {
            System.err.println("Usage: WordCountMapReduce <input_path> <output_path> <word_to_count>");
            System.exit(-1);
        }

        Configuration conf = new Configuration();
        conf.set("targetWord", args[2]); // Pasamos la palabra objetivo como parámetro

        Job job = Job.getInstance(conf, "Word Count for Specific Word");
        job.setJarByClass(wordcount_mapreduce.class);
        job.setMapperClass(WordCountMapper.class);
        // Elimina la línea que asigna el Combiner
        job.setReducerClass(WordCountReducer.class);
        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(IntWritable.class);

        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileOutputFormat.setOutputPath(job, new Path(args[1]));

        System.exit(job.waitForCompletion(true) ? 0 : 1);
    }
}



Writing wordcount_mapreduce.java


Compilamos el archivo .java para generar .jar y poder ejecutarlo.

In [26]:
%%bash
javac -cp "/usr/local/hadoop-3.3.6/share/hadoop/common/*:/usr/local/hadoop-3.3.6/share/hadoop/mapreduce/*:/usr/local/hadoop-3.3.6/share/hadoop/common/lib/*" wordcount_mapreduce.java -Xlint
jar -cvf wordcountmapreduce.jar *.class

added manifest
adding: wordcount_mapreduce$WordCountMapper.class(in = 3311) (out= 1353)(deflated 59%)
adding: wordcount_mapreduce$WordCountReducer.class(in = 1791) (out= 754)(deflated 57%)
adding: wordcount_mapreduce.class(in = 1813) (out= 971)(deflated 46%)


4 warnings


In [27]:
!hadoop jar wordcountmapreduce.jar wordcount_mapreduce /dataset /output_apple apple

2024-12-19 11:57:50,287 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2024-12-19 11:57:50,496 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2024-12-19 11:57:50,496 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2024-12-19 11:57:50,634 WARN mapreduce.JobResourceUploader: Hadoop command-line option parsing not performed. Implement the Tool interface and execute your application with ToolRunner to remedy this.
2024-12-19 11:57:50,786 INFO input.FileInputFormat: Total input files to process : 2
2024-12-19 11:57:50,832 INFO mapreduce.JobSubmitter: number of splits:2
2024-12-19 11:57:51,180 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local976075796_0001
2024-12-19 11:57:51,180 INFO mapreduce.JobSubmitter: Executing with tokens: []
2024-12-19 11:57:51,461 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
2024-12-19 11:57:51,462 INFO mapreduce.Job: Running job: job_local976075796_00

In [28]:
!hdfs dfs -ls /dataset

Found 2 items
-rw-r--r--   1 root root   14322766 2024-12-19 11:57 /dataset/data.json
-rw-r--r--   1 root root    1082761 2024-12-19 11:57 /dataset/processed_data2.csv


In [29]:
!hdfs dfs -cat /output_apple/part*

data.json : apple	453
processed_data2.csv : apple	1164


# MapReduce Python con Hadoop-Streaming

In [30]:
!rm -r output_python
!rm -r map_word_in_file.py
!rm -r reducer_word_in_file.py

rm: cannot remove 'output_python': No such file or directory
rm: cannot remove 'map_word_in_file.py': No such file or directory
rm: cannot remove 'reducer_word_in_file.py': No such file or directory


Creamos el archivo para el map

In [31]:
%%writefile -a map_word_in_file.py

import sys
import os
import re

# La palabra a buscar se pasa como argumento
target_word = sys.argv[1].lower()

# Leer líneas de la entrada estándar
for line in sys.stdin:
    # Obtener el nombre del fichero desde las variables de entorno de Hadoop
    filename = os.environ.get('map_input_file', 'unknown_file')

    # Limpiar y normalizar la línea
    line = line.strip().lower()

    # Usar una expresión regular para encontrar coincidencias exactas de palabras
    pattern = r'\b{}\b'.format(re.escape(target_word))  # \b asegura coincidencias completas
    matches = re.findall(pattern, line)

    # Si hay coincidencias, contar y emitir
    if matches:
        count = len(matches)
        print(f"{filename}\t{count}")


Writing map_word_in_file.py


Creamos el archivo para el reducer

In [32]:
%%writefile -a reducer_word_in_file.py

import sys

current_file = None
current_count = 0

import sys

current_file = None
current_count = 0

# Leer líneas de la entrada estándar
for line in sys.stdin:
    # Dividir la entrada en filename y count
    filename, count = line.strip().split('\t', 1)
    try:
        count = int(count)
    except ValueError:
        continue

    # Acumular el conteo para cada archivo
    if current_file == filename:
        current_count += count
    else:
        if current_file:
            # Emitir el resultado para el archivo anterior
            print(f"{current_file}\t{current_count}")
        # Reiniciar el acumulador para el nuevo archivo
        current_file = filename
        current_count = count

# Emitir el resultado del último archivo
if current_file:
    print(f"{current_file}\t{current_count}")

Writing reducer_word_in_file.py


Damos permisos a los archivos creados

In [33]:
!chmod +x map_word_in_file.py reducer_word_in_file.py
!ls -la

total 16456
drwxr-xr-x 1 root root     4096 Dec 19 11:57  .
drwxr-xr-x 1 root root     4096 Dec 19 11:57  ..
drwxr-xr-x 4 root root     4096 Dec 17 14:21  .config
-rw-r--r-- 1 root root 14322766 Oct 28 16:48  data.json
drwxr-xr-x 2 root root     4096 Dec 19 11:57  grep_apple
-rwxr-xr-x 1 root root      731 Dec 19 11:57  map_word_in_file.py
-rw-r--r-- 1 root root  1082761 Oct 28 16:48  processed_data2.csv
-rwxr-xr-x 1 root root      816 Dec 19 11:57  reducer_word_in_file.py
drwxr-xr-x 1 root root     4096 Dec 17 14:21  sample_data
drwxr-xr-x 2 root root     4096 Dec 19 11:57  samsung
-rw-r--r-- 1 root root  1385178 Dec 19 11:56  telefonos-2024.zip
-rw-r--r-- 1 root root     3311 Dec 19 11:57 'wordcount_mapreduce$WordCountMapper.class'
-rw-r--r-- 1 root root     1791 Dec 19 11:57 'wordcount_mapreduce$WordCountReducer.class'
-rw-r--r-- 1 root root     1813 Dec 19 11:57  wordcount_mapreduce.class
-rw-r--r-- 1 root root     3898 Dec 19 11:57  wordcountmapreduce.jar
-rw-r--r-- 1 root root   

Ejecutamos el map y el reducer utilizando Hadoop-Sreaming

In [34]:
%%bash
hadoop jar /usr/local/hadoop-3.3.6/share/hadoop/tools/lib/hadoop-streaming-3.3.6.jar \
   -input /dataset \
   -output output_python \
   -file map_word_in_file.py  -file reducer_word_in_file.py  \
   -mapper 'python map_word_in_file.py apple'  -reducer 'python reducer_word_in_file.py'

packageJobJar: [map_word_in_file.py, reducer_word_in_file.py] [] /tmp/streamjob14994178510656868071.jar tmpDir=null


2024-12-19 11:58:00,868 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
2024-12-19 11:58:01,610 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2024-12-19 11:58:01,783 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2024-12-19 11:58:01,783 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2024-12-19 11:58:01,807 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2024-12-19 11:58:02,063 INFO mapred.FileInputFormat: Total input files to process : 2
2024-12-19 11:58:02,090 INFO mapreduce.JobSubmitter: number of splits:2
2024-12-19 11:58:02,409 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local1340815619_0001
2024-12-19 11:58:02,409 INFO mapreduce.JobSubmitter: Executing with tokens: []
2024-12-19 11:58:02,871 INFO mapred.LocalDistributedCacheManager: Localized file:/content/map_word_in_file.py as file:/tmp/hadoop-root/mapred/local/j

Comprobamos los resultados

In [35]:
!ls -l output_python

total 4
-rw-r--r-- 1 root root 67 Dec 19 11:58 part-00000
-rw-r--r-- 1 root root  0 Dec 19 11:58 _SUCCESS


In [36]:
!cat ./output_python/part-00000

file:/dataset/data.json	453
file:/dataset/processed_data2.csv	1164
